## 1.3 Computer arithmetic

The way computers store numbers and perform computations could surprise
the beginner. In Python if you type $(\sqrt{3})^{2}$ the result will
be 2.9....96, where 9 is repeated 15 times. Here are two obvious
but fundamental differences in the way computers do arithmetic:

- only finitely many numbers can be represented in a computer;
- a number represented in a computer can only have finitely many digits.

Therefore the numbers that can be represented in a computer exactly
is only a subset of rational numbers. Anytime the computer performs
an operation whose outcome is not a number that can be represented
exactly in the computer, an approximation will replace the exact number.
This is called the *roundoff error*: error produced when a computer
is used to perform real number calculations.

### Floating-point representation of real numbers

Here is a general model for representing real numbers in a computer:

\begin{equation}\label{eq:float pt general}
x=s(.a_{1}a_{2}...a_{t})_{\beta}\times\beta^{e}
\end{equation}

where

\begin{align*}
s & \rightarrow\text{sign of } x=\pm 1\\
e & \rightarrow\text{exponent, with bounds } L\le e\leq U\\
(.a_{1}...a_{t})_{\beta} & =\frac{a_{1}}{\beta}+\frac{a_{2}}{\beta^{2}}+...+\frac{a_{t}}{\beta^{t}};\text{ the mantissa}\\
\beta & \rightarrow\text{base}\\
t & \rightarrow\text{number of digits; the precision.}
\end{align*}

In the floating-point representation \eqref{eq:float pt general},
if we specify $e$ in such a way that $a_{1}\neq 0,$ then the representation
will be unique. This is called the **normalized** floating-point
representation. For example if $\beta=10,$ in the normalized floating-point
we would write $0.012$ as $0.12\times10^{-1}$, instead of choices
like $0.012 \times 10^{0}$ or $0.0012 \times 10$.

In most computers today, the base is $\beta=2.$ Bases 8 and 16 were
used in old IBM mainframes in the past. Some handheld calculators
use base 10. An interesting historical example is a short-lived
computer named Setun developed at Moscow State University which used base 3.

There are several choices to make in the general floating-point model
\eqref{eq:float pt general} for the values of $s,\beta, t,e$. The IEEE 64-bit floating-point representation 
is the specific model used in most computers today:

\begin{equation}\label{eq:IEEE}
x=(-1)^{s}(1.a_{2}a_{3}...a_{53})_{2}2^{e-1023}.
\end{equation}

Some comments:

- Notice how $s$ appears in different forms in equations \eqref{eq:float pt general} and \eqref{eq:IEEE}. In \eqref{eq:IEEE}, $s$ is either 0 or 1. If $s=0$, then $x$ is positive. If $s=1$, $x$ is negative.

- Since $\beta=2,$ in the normalized floating-point representation of $x$ the first (nonzero) digit after the decimal point has to be 1. Then we do not have to store this number. That's why we write $x$ as a decimal number starting at 1 in \eqref{eq:IEEE}. Even though precision is $t=52,$ we are able to access up to the 53rd digit $a_{53}$.

- The bounds for the exponent are: $0\leq e\leq 2047.$ We will discuss where 2047 comes from shortly. But first, let's discuss why we have $e-1023$ as the exponent in the representation \eqref{eq:IEEE}, as opposed to simply $e$ (which we had in the representation \eqref{eq:float pt general}). If the smallest exponent possible was $e=0$, then the smallest positive number the computer can generate would be $(1.00...0)_{2}=1$: certainly we need the computer to represent numbers less than 1! That's why we use the shifted expression $e-1023$, called the \textbf{biased exponent},  in the representation In \eqref{eq:IEEE}. Note that the bounds for the biased exponent are $-1023\leq e-1023 \leq 1024.$

Here is a schema that illustrates how the physical bits of a computer
correspond to the representation above. Each cell in the table below,
numbered 1 through 64, correspond to the physical bits in the computer
memory.

\begin{equation*}
\begin{array}{|c|c|c|c|c|c|c|c|}
\hline
1 & \color{blue}2 & \color{blue}{3} & \color{blue}{\ldots} & \color{blue}{12} & \color{red}{13} & \color{red}{\ldots} & \color{red}{64}\\
\hline
\end{array}
\end{equation*}

- The first bit is the sign bit: it stores the value for $s$, 0 or 1.
- The blue bits 2 through 12 store the exponent $e$ (not $e-1023)$. Using 11 bits, one can generate the integers from 0 to $2^{11}-1=2047$. Here is how you get the smallest and largest values for $e$:
\begin{align*}
e&=(00...0)_{2}=0\\
e&=(11...1)_{2}=2^{0}+2^{1}+...+2^{10}=\frac{2^{11}-1}{2-1}=2047.
\end{align*}
- The red bits, and there are 52 of them, store the digits $a_{2}$ through $a_{53}$.

### Example
Find the floating-point representation of 10.375.

**Solution.**

You can check that $10=({\color{cyan}{\color{red}1}}{\color{brown}{\color{green}{\color{green}0}}}{\color{blue}1}{\color{black}0})_{2}$
and $0.375=(.0{\color{blue}1}{\color{green}1})_{2}$ by computing
\begin{align*}
10 & ={\color{black}0}\times2^{0}+{\color{blue}1}\times2^{1}+{\color{brown}{\color{green}{\color{green}0}}}\times2^{2}+{\color{cyan}{\color{red}1}}\times2^{3}\\
0.375 & =0\times2^{-1}+{\color{blue}1}\times2^{-2}+{\color{green}1}\times2^{-3}.
\end{align*}
Then
\begin{equation*}
10.375=(1010.011)_{2}=(1.010011)_{2}\times2^{3}
\end{equation*}
where $(1.010011)_{2}\times2^{3}$ is the normalized floating-point representation of the number.
Now we rewrite this in terms of the representation \eqref{eq:IEEE}:
\begin{equation*}
10.375=(-1)^{0}(1.010011)_{2}\times2^{1026-1023}.
\end{equation*}
Since $1026=(10000000010)_{2}$, the bit by bit representation is:

\begin{equation*}
\begin{array}{|c||c|c|c|c|c|c|c|c|c|c|c||c|c|c|c|c|c|c|c|c|}
\hline
0 & \color{blue}{1} & \color{blue}{0} & \color{blue}{0} & \color{blue}{0} & \color{blue}{0} & \color{blue}{0} & \color{blue}{0} & \color{blue}{0} & \color{blue}{0} & \color{blue}{1} & \color{blue}{0} & \color{red}{0} & \color{red}{1} & \color{red}{0} & \color{red}{0} & \color{red}{1} & \color{red}{1} & 0 & \ldots & 0\\
\hline
\end{array}
\end{equation*}

Notice the first sign bit is 0 since the number is positive. The next
11 bits (in blue) represent the exponent $e=1026$, and the next group of
red bits are the mantissa, filled with 0's after the last digit of
the mantissa. In Python, although there is no built-in function that produces
the bit by bit representation of a number, we can define the following function named *float2bin*, which provides the bit representation of a floating-point number, based on the *struct* package:

In [1]:
import struct
def float2bin(f):
    (d,) = struct.unpack(">Q", struct.pack(">d", f))
    return f'{d:064b}'

In [2]:
float2bin(10.375)

'0100000000100100110000000000000000000000000000000000000000000000'

#### Special cases: zero, infinity, NAN

In the floating-point arithmetic there are two zeros: $+0.0$ and $-0.0$, and they have special representations. In the representation of zero, all exponent and mantissa bits are set to 0. The sign bit is 0 for $+0.0$, and 1, for $-0.0$:
\begin{equation*}
0.0\rightarrow
\begin{array}{|c|c|c|c|c|c|c|}
\hline
0 & \color{blue}{0} & \color{blue}{\text{all zeros}} & \color{blue}{0} & \color{red}{0} & \color{red}{\text{all zeros}} & \color{red}{0}\\
\hline
\end{array}
\end{equation*}\

\begin{equation*}
-0.0\rightarrow 
\begin{array}{|c|c|c|c|c|c|c|}
\hline
1 & \color{blue}{0} & \color{blue}{\text{all zeros}} & \color{blue}{0} & \color{red}{0} & \color{red}{\text{all zeros}} & \color{red}{0}\\
\hline
\end{array}
\end{equation*}

When the exponent bits are set to zero, we have $e=0$ and thus $e-1023=-1023.$
This arrangement, all exponent bits set to zero, is reserved for
$\pm 0.0$ and **subnormal numbers**. Subnormal
numbers are an exception to our normalized floating-point representation,
an exception that is useful in various ways. For details see *Goldberg (1991)*.

Here is how plus and minus infinity is represented in the computer:
\begin{equation*}
\infty\rightarrow
\begin{array}{|c|c|c|c|c|c|c|}
\hline
0 & \color{blue}{1} & \color{blue}{\text{all ones}} & \color{blue}{1} & \color{red}{0} & \color{red}{\text{all zeros}} & \color{red}{0}\\
\hline
\end{array}
\end{equation*}\

\begin{equation*}
-\infty\rightarrow 
\begin{array}{|c|c|c|c|c|c|c|}
\hline
1 & \color{blue}{1} & \color{blue}{\text{all ones}} & \color{blue}{1} & \color{red}{0} & \color{red}{\text{all zeros}} & \color{red}{0}\\
\hline
\end{array}
\end{equation*}

When the exponent bits are set to one, we have $e=2047$ and thus
$e-1023=1024.$ This arrangement is reserved for $\pm \infty$ as
well as other special values such as NaN (not-a-number).

In conclusion, even though $-1023\leq e-1023\leq 1024$,
when it comes to representing non-zero real numbers, we only have
access to exponents in the following range: $-1022\leq e-1023\leq 1023.$

Therefore, the smallest positive real number that can be represented in a computer is
\begin{equation*}
x=(-1)^{0}(1.00\ldots0)_{2}\times2^{-1022}=2^{-1022}\approx0.2\times10^{-307}
\end{equation*}
and the largest is
\begin{align*}
x=(-1)^{0}(1.11\ldots1)_{2}\times2^{1023}&=\left(1+\frac{1}{2}+\frac{1}{2^{2}}+...+\frac{1}{2^{52}}\right)\times2^{1023}\\
&=(2-2^{-52})2^{1023}\\
&\approx0.18\times10^{309}.
\end{align*}

During a calculation, if a number
less than the smallest floating-point number is obtained, then we obtain
the **underflow error**. A number greater than the largest
gives **overflow error**.

### Exercise 1.3-1

Consider the following toy model for a normalized floating-point representation
in base 2: $x=(-1)^{s}(1.a_{2}a_{3})_{2}\times2^{e}$ where $-1\leq e\leq1$.
Find all positive machine numbers (there are 12 of them) that can
be represented in this model. Convert the numbers to base 10, and
then carefully plot them on the number line, by hand, and comment
on how the numbers are spaced.

### Representation of integers

In the previous section, we discussed representing real numbers in a computer. Here we will give a brief discussion of representing integers. How does a computer represent an integer $n$? As in real numbers, we start with writing $n$ in base 2. We have 64 bits to represent its digits and sign. As in the floating-point representation, we can allocate one bit for the sign, and use the rest, 63 bits, for the digits. This approach has some disadvantages when we start adding integers. Another approach, known as the **two's complement**, is more commonly used, including in Python.

For an example, assume we have 8 bits in our computer. To represent 12 in two's complement (or any positive integer), we simply write it in its base 2 expansion: $(00001100)_2$. To represent $-12$, we do the following: flip all digits, replacing 1 by 0, and 0 by 1, and then add 1 to the result. When we flip digits for 12, we get  $(11110011)_2$, and adding 1 (in binary), gives $(11110100)_2$. Therefore $-12$ is represented as $(11110100)_2$ in two's complement approach. It may seem mysterious to go through all this trouble to represent $-12$, until you add the representations of 12 and $-12$,
\begin{equation*}
(00001100)_2 + (11110100)_2 = (\color{red}{1}00000000)_2
\end{equation*}
and realize that the first 8 digits of the sum (from right to left), which is what the computer can only represent (ignoring the red digit 1), is $(00000000)_2$. So just like $12+(-12)=0$ in base 10, the sum of the representations of these numbers is also 0.

We can repeat these calculations with 64-bits, using Python. The function *int2bin* defined below outputs the digits of an integer, using two's complement for negative numbers:

In [3]:
import struct
def int2bin(i):
    (d,) = struct.unpack(">Q", struct.pack(">q", i))    
    return f'{d:064b}'

In [4]:
int2bin(12)

'0000000000000000000000000000000000000000000000000000000000001100'

```{bibliography}
```

In [5]:
int2bin(-12)

'1111111111111111111111111111111111111111111111111111111111110100'

You can verify that the sum of these representations is 0, when truncated to 64-digits.

Here is another example illustrating the advantages of two's complement. Consider $-3$ and 5, with representations,
\begin{equation*}
-3=(11111101)_2 \text{ and }
5=(00000101)_2.
\end{equation*}

The sum of $-3$ and 5 is 2; what about the binary sum of their representations? We have
\begin{equation*}
(11111101)_2+(00000101)_2=(\color{red}{1}00000010)_2
\end{equation*}
and if we ignore the ninth bit in red, the result is $(10)_2$, which is indeed 2. Notice that if we followed the same approach used in the floating-point representation and allocated the leftmost bit to the sign of the integer, we would not have had this property.

We will not discuss integer representations and integer arithmetic any further. However one useful fact to keep in mind is the following: in two's complement, using 64 bits, one can represent integers between $-2^{63}=-9223372036854775808$ and $2^{63}-1=9223372036854775807$. Any integer below or above yields **underflow** or **overflow** error. However, Python is special compared to the other programming languages in that it supports arbitrary-precision integer implementation. Integers of any size can be supported in Python, while it is not true for floats.

### Example
From Calculus, we know that $\lim_{n->\infty} \frac{n^n}{n!}=\infty$. Therefore, computing $\frac{n^n}{n!}$, which is a float, for large $n$ will cause overflow at some point. Here is a Python code for this calculation:

In [6]:
from scipy.special import factorial
f = lambda n: n**n/factorial(n)

Let's have a closer look at this function. The Python function \textbf{factorial(n)} computes the factorial, returned as a float, as the gamma function at $n+1$. If we call $f$ with an integer input, then the above code will compute $n^n$ as an integer and $\text{factorial}(n)$ as a float. Then it will divide these numbers, by first converting $n^n$ to a float, and obtain a floating-point factorial. Let's compute
$f(n)=\frac{n^n}{n!}$ as $n=1,...,1000$:

In [7]:
for n in range(1, 1001):
    print(n, f(n))

1 1.0
2 2.0
3 4.5
4 10.666666666666666
5 26.041666666666668
6 64.8
7 163.4013888888889
8 416.1015873015873
9 1067.6270089285715
10 2755.731922398589
11 7147.658895778219
12 18613.926233766233
13 48638.8461384701
14 127463.00337621226
15 334864.627690599
16 881657.9515664614
17 2325750.6353670303
18 6145596.911985091
19 16263866.353935543
20 43099804.12182177
21 114356611.3686037
22 303761260.04766583
23 807692034.6582023
24 2149632061.4646716
25 5726042115.469875
26 15264691107.046675
27 40722913454.160416
28 108713644516.20932
29 290404350962.9901
30 776207020879.728
31 2075825009901.003
32 5554277310717.434
33 14868745536619.768
34 39821588060720.56
35 106696188125885.62
36 285992654335445.2
37 766879127067901.0
38 2057103777254334.5
39 5519943051629867.0
40 1.4816805671322662e+16
41 3.9784069108519944e+16
42 1.0685429454599882e+17
43 2.8707602164766144e+17
44 7.714688909178948e+17
45 2.073725455492908e+18
46 5.575586585767318e+18
47 1.4994492532091273e+19
48 4.03339361002671e+19
49 

OverflowError: int too large to convert to float

Notice that the process cannot proceed beyond $n=143$.  Where exactly does the error occur? Python can compute $144^{144}$ exactly (arbitrary-precision); however, when it is converted to a floating-point number, overflow occurs. For $143^{143}$, overflow does not happen.

In [ ]:
float(143**143)

In [ ]:
float(144**144)

The function $f(n)$ can be coded in a much better way if we rewrite $\frac{n^n}{n!}$ as $\frac{n}{n}\frac{n}{n-1}...\frac{n}{1}$. Each fraction can be computed separately, and then multiplied,  which will slow down the growth of the numbers. Here is a new code using a \textbf{for} statement.

In [ ]:
def f(n):
    pr = 1.
    for i in range(1, n):
        pr *= n/(n-i)
    return pr

Let's compute $f(n)=\frac{n^n}{n!}$ as $n=1,...,1000$ again:

In [ ]:
for i in range(1, 1001):
    print(i, f(i))

The previous version of the code gave overflow error when $n=144$. This version has no difficulty in computing $n^n/n!$ for $n=144$. In fact, we can go as high as $n=713$. Overflow in floating-point arithmetic yields the output inf, which stands for infinity.

Another way to accommodate a larger value of $n$ is to define the function $f$ in this way:

In [ ]:
from scipy.special import factorial
f = lambda n: n**n/factorial(n, True)

With the addition input "True'', the function **factorial(n)** returns an integer with arbitrary precision, instead of a float, so that both the numerator and denominator are integers. The limit process stops only when the division produces a floating-point number that is so large that overflow occurs. As a result, the same result as in the improved algorithm above is expected. Please verify that the process can continue until $n=713$ with the modified $f$ function.

We will discuss several features of computer arithmetic in the rest
of this section. The discussion is easier to follow if we use the familiar
base 10 representation of numbers instead of base 2. To this end, we introduce the **normalized
decimal floating-point representation**:
\begin{equation*}
\pm0.d_{1}d_{2}...d_{k}\times10^{n}
\end{equation*}
where $1\leq d_{1}\leq9,0\leq d_{i}\leq9$ for all $i=2,3,...,k.$
Informally, we call these numbers $k-$digit decimal machine numbers.


### Chopping \& Rounding

Let $x$ be a real number with more digits the computer can handle:
$x=0.d_{1}d_{2}\ldots d_{k}d_{k+1}\ldots\times10^{n}$. How will the
computer represent $x$? Let's use the notation $fl(x)$ for the floating-point
representation of $x$. There are two choices, chopping and rounding:
- In chopping, we simply take the first $k$ digits and ignore the rest: $fl(x)=0.d_{1}d_{2}\ldots d_{k}\times10^{n}.$
- In rounding, if $d_{k+1}\geq 5$ we add 1 to $d_{k}$ to obtain $fl(x)$. If $d_{k+1}<5,$ then we simply do as in chopping.

### Example
Find 5-digit ($k=5$) chopping and rounding values of the numbers
below:
- $\pi=0.314159265...\times10^{1}$

  Chopping gives $fl(\pi)=0.31415\times10^{1}$ and rounding gives $fl(\pi)=0.31416\times10^{1}.$
- $0.0001234567$

  We need to write the number in the normalized representation first as $0.1234567\times10^{-3}.$ Now chopping gives $0.12345\times10^{-3}$ and rounding gives $0.12346\times10^{-3}$.

### Absolute and relative error

Since computers only give approximations to real numbers, we need to be clear on how we measure the error of an approximation.

### Definition
Suppose $x^{*}$ is an approximation to $x$.
- $|x^{*}-x|$ is called the **absolute error**
- $\frac{|x^{*}-x|}{|x|}$ is called the **relative error** $(x\neq0)$
\end{definition}

Relative error usually is a better choice of measure, and we need
to understand why.

### Example

Find absolute and relative errors of
1. $x=0.20\times10^{1},x^{*}=0.21\times10^{1}$
2. $x=0.20\times10^{-2},x^{*}=0.21\times10^{-2}$
3. $x=0.20\times10^{5},x^{*}=0.21\times10^{5}$

Notice the only difference in the three cases is the exponent
of the numbers. The absolute errors are: $0.01\times 10$ , $0.01\times 10^{-2}$
, $0.01\times 10^{5}$. The absolute errors are different since the
exponents are different. However, the relative error in each case
is the same: $0.05$.

### Definition
The number $x^{*}$ is said to approximate $x$ to $s$ significant
digits (or figures) if $s$ is the largest nonnegative integer such
that
\begin{equation*}
\frac{|x-x^{*}|}{|x|}\leq5\times10^{-s}.
\end{equation*}

In Example above we had $\frac{|x-x^{*}|}{|x|}=0.05\leq5\times10^{-2}$
but not less than or equal to $5\times10^{-3}$. Therefore we say
$x^{*}=0.21$ approximates $x=0.20$ to 2 significant digits (but not to 3 digits).

When the computer approximates a real number $x$ by $fl(x)$, what
can we say about the error? The following result gives an upper bound
for the relative error.

### Lemma
The relative error of approximating
$x$ by $fl(x)$ in the $k$-digit normalized decimal floating-point representation satisfies
\begin{equation*}
\frac{|x-fl(x)|}{|x|}\leq\begin{cases}
10^{-k+1} & \text{if chopping}\\
\frac{1}{2}(10^{-k+1}) & \text{if rounding.}
\end{cases}
\end{equation*}

### Proof
We will give the proof for chopping; the proof for rounding is similar
but tedious. Let
\begin{equation*}
x=0.d_{1}d_{2}...d_{k}d_{k+1}...\times 10^{n}.
\end{equation*}
Then
\begin{equation*}
fl(x)=0.d_{1}d_{2}...d_{k}\times 10^{n}
\end{equation*}
if chopping is used. Observe that
\begin{equation*}
\frac{|x-fl(x)|}{|x|}=\frac{0.d_{k+1}d_{k+2}...\times 10^{n-k}}{0.d_{1}d_{2}...\times 10^{n}}=\left(\frac{0.d_{k+1}d_{k+2}...}{0.d_{1}d_{2}...}\right)10^{-k}.
\end{equation*}
We have two simple bounds: $0.d_{k+1}d_{k+2}...<1$ and $0.d_{1}d_{2}...\geq 0.1$,
the latter true since the smallest value $d_{1}$ can be, is 1. Using these
bounds in the equation above we get
\begin{equation*}
\frac{|x-fl(x)|}{|x|}\leq\frac{1}{0.1}10^{-k}=10^{-k+1}.
\end{equation*}



### Remark
Lemma \ref{lem:relative-error-chopping} easily transfers to the base
2 floating-point representation: $x=(-1)^{s}(1.a_{2}...a_{53})_{2}\times2^{e-1023}$, by
\begin{equation*}
\frac{|x-fl(x)|}{|x|}\leq\begin{cases}
2^{-t+1}=2^{-53+1}=2^{-52} & \text{if chopping}\\
\frac{1}{2}(2^{-t+1})=2^{-53} & \text{if rounding}.
\end{cases}
\end{equation*}

### Machine epsilon

Machine epsilon $\epsilon$ is the smallest positive floating point
number for which $fl(1+\epsilon)>1.$ This means, if we add to $1.0$
any number less than $\epsilon$, the machine computes the sum as
$1.0$.

The number 1.0 in its binary floating-point representation is simply
$(1.0\ldots0)_{2}$ where $a_{2}=a_{3}=...=a_{53}=0.$ We want to
find the smallest number that gives a sum larger than 1.0, when it
is added to 1.0. The answer depends on whether we chop or round.

If we are chopping, examine the binary addition

\begin{equation*}
\begin{array}{cccccc}
 &  & a_{2} &  & a_{52} & a_{53}\\
 & 1. & 0 & ... & 0 & 0\\
+ & 0. & 0 & ... & 0 & 1\\
\hline
 & 1. & 0 & ... & 0 & 1
\end{array}
\end{equation*}

and notice $(0.0...01)_{2}=\left(\frac{1}{2}\right)^{52}=2^{-52}$
is the smallest number we can add to 1.0 such that the sum will be
different than 1.0.

If we are rounding, examine the binary addition

\begin{equation*}
\begin{array}{cccccc}
 &  & a_{2} &  & a_{52} & a_{53} & \\
 & 1. & 0 & ... & 0 & 0 & \\
+ & 0. & 0 & ... & 0 & 0 & 1\\
\hline
 & 1. & 0 & ... & 0 & 0 & 1\\
\end{array}
\end{equation*}

where the sum has to be rounded to 53 digits to obtain

\begin{equation*}
\begin{array}{ccccc}
 & a_{2} &  & a_{52} & a_{53}\\
1. & 0 & ... & 0 & 1\\
\end{array}
\end{equation*}

Observe that the number added to $1.0$ above is $(0.0...01)_{2}=\left(\frac{1}{2}\right)^{53}=2^{-53}$,
which is the smallest number that will make the sum larger than 1.0 with rounding.

In summary, we have shown
\begin{equation*}
\epsilon=\begin{cases}
2^{-52} & \text{if chopping}\\
2^{-53} & \text{if rounding}
\end{cases}.
\end{equation*}
As a consequence, notice that we can restate the inequality in Remark \ref{rem:rem_base2relerror} in a compact way using the machine epsilon as:
\begin{equation*}
\frac{|x-fl(x)|}{|x|}\leq \epsilon.
\end{equation*}


\begin{remark}
There is another definition of machine epsilon: it is the distance
between 1.0 and the next floating-point number.

\begin{equation*}
\begin{array}{cccccc}
&  & a_{2} &  & a_{52} & a_{53}\\
\text{number } 1.0 & 1. & 0 & ... & 0 & 0\\
\text{next number }& 1. & 0 & ... & 0 & 1\\
\hline
\text{distance} & 0. & 0 & ... & 0 & 1
\end{array}
\end{equation*}

Note that the distance (absolute value of the difference) is $\left(\frac{1}{2}\right)^{52}=2^{-52}.$
In this alternative definition, machine epsilon is not based on whether
rounding or chopping is used. Also, note that the distance between
two adjacent floating-point numbers is not constant, but it is smaller
for smaller values, and larger for larger values (see Exercise \ref{ex:1.3-1}).
\end{remark}

### Propagation of error

We discussed the resulting error when chopping or rounding is used to approximate
a real number by its machine version. Now imagine carrying out a long
calculation with many arithmetical operations, and at each step there
is some error due to say, rounding. Would all the rounding errors
accumulate and cause havoc? This is a rather difficult question to
answer in general. For a much simpler example, consider adding two real numbers
$x,y$. In the computer, the numbers are represented as $fl(x),fl(y)$.
The sum of these number is $fl(x)+fl(y)$; however, the computer can only represent its floating-point version, $fl(fl(x)+fl(y))$.
Therefore the relative error in adding two numbers is:
\begin{equation*}
\left|\frac{(x+y)-fl(fl(x)+fl(y))}{x+y}\right|.
\end{equation*}
In this section, we will look at some specific examples where
roundoff error can cause problems, and how we can avoid them.

#### Subtraction of nearly equal quantities: Cancellation of leading digits

The best way to explain this phenomenon is by an example. Let $x=1.123456,y=1.123447$.
We will compute $x-y$ and the resulting roundoff error using rounding
and 6-digit arithmetic. First, we find $fl(x),fl(y)$:
\begin{equation*}
fl(x)=1.12346,fl(y)=1.12345.
\end{equation*}
The absolute and relative error due to rounding is:
\begin{align*}
|x-fl(x)| & =4\times10^{-6},|y-fl(y)|=3\times 10^{-6}\\
\frac{|x-fl(x)|}{|x|} & =3.56\times10^{-6},\frac{|y-fl(y)|}{|y|}=2.67\times 10^{-6}.
\end{align*}
From the relative errors, we see that $fl(x)$ and $fl(y)$ approximate $x$
and $y$ to six significant digits. Let's see how the error propagates
when we subtract $x$ and $y$. The actual difference is:
\begin{equation*}
x-y=1.123456-1.123447=0.000009=9\times 10^{-6}.
\end{equation*}
The computer finds this difference first by computing $fl(x),fl(y),$
then taking their difference and approximating the difference by its
floating-point representation: $fl(fl(x)-fl(y))$:
\begin{equation*}
fl(fl(x)-fl(y))=fl\left(1.12346-1.12345\right)=10^{-5}.
\end{equation*}
The resulting absolute and relative errors are:
\begin{align*}
\left|(x-y)-(fl(fl(x)-fl(y)))\right| & =10^{-6}\\
\frac{|(x-y)-(fl(fl(x)-fl(y)))|}{|x-y|} & =0.1.
\end{align*}
Notice how large the relative error is compared to the absolute error!
The machine version of $x-y$ approximates $x-y$ to only one significant digit. Why did this
happen? When we subtract two numbers that are nearly equal, the leading
digits of the numbers cancel, leaving a result close to the rounding
error. In other words, the rounding error dominates the difference.

#### Division by a small number

Let $x=0.444446$ and compute $\frac{x}{10^{-5}}$ in a computer with
5-digit arithmetic and rounding. We have $fl(x)=0.44445$, with an
absolute error of $4\times10^{-6}$ and relative error of $9\times10^{-6}$.
The exact division is $\frac{x}{10^{-5}}=0.444446\times10^{5}$. The
computer computes: $fl\left(\frac{x}{10^{-5}}\right)=0.44445\times10^{5}$,
which has an absolute error of $0.4$ and relative error of $9\times10^{-6}$.
The absolute error went from $4\times10^{-6}$ to $0.4$. Perhaps
not surprisingly, division by a small number magnifies the absolute
error but not the relative error.

Consider the computation of
\begin{equation*}
\frac{1-\cos x}{\sin x}
\end{equation*}
when $x$ is near zero. This is a problem where we have both \textit{subtraction of nearly equal quantities} which happens in the numerator, and \textit{division by a small number}, when $x$ is close to zero. Let $x=0.1$. Continuing with five-digit rounding, we have
\begin{align*}
&fl(\sin 0.1)=0.099833, fl(\cos 0.1)=0.99500\\
&fl\left(\frac{1-\cos 0.1}{\sin 0.1}\right)=0.050084.
\end{align*}
The exact result to 8 digits is 0.050041708, and the relative error of this computation is $8.5\times 10^{-4}$. Next we will see how to reduce this error using a simple algebraic identity.

#### Ways to avoid loss of accuracy

Here we will discuss some examples where a careful rewriting of the
expression to compute can make the roundoff error much smaller.

### Example
Let's revisit the calculation of
\begin{equation*}
\frac{1-\cos x}{\sin x}.
\end{equation*}
Observe that using the algebraic identity
\begin{equation*}
\frac{1-\cos x}{\sin x}=\frac{\sin x}{1+\cos x}
\end{equation*}
removes both difficulties encountered before: there is no cancellation of significant digits or division by a small number. Using five-digit rounding, we have
\begin{equation*}
fl\left(\frac{\sin 0.1}{1+\cos 0.1}\right)=0.050042.
\end{equation*}
The relative error is $5.8\times 10^{-6}$, about a factor of 100 smaller than the error in the original computation.

### Example
Consider the quadratic formula: the solution of $ax^{2}+bx+c=0$ is

\begin{equation*}
r_{1}=\frac{-b+\sqrt{b^{2}-4ac}}{2a},r_{2}=\frac{-b-\sqrt{b^{2}-4ac}}{2a}.
\end{equation*}
If $|b|\approx\sqrt{b^{2}-4ac}$, then we have a potential loss of precision in computing one of the roots
due to cancellation. Let's consider a specific equation: $x^{2}-11x+1=0$.
The roots from the quadratic formula are: $r_{1}=\frac{11+\sqrt{117}}{2}\approx 10.90832691$,
and $r_{2}=\frac{11-\sqrt{117}}{2}\approx 0.09167308680.$

Next will use four-digit
arithmetic with rounding to compute the roots:
\begin{gather*}
fl(\sqrt{117})=10.82\\
fl(r_{1})=fl\left(\frac{fl(fl(11.0)+fl(\sqrt{117}))}{fl(2.0)}\right)=fl\left(\frac{fl(11.0+10.82)}{2.0}\right)=
fl\left(\frac{21.82}{2.0}\right)=10.91\\
fl(r_{2})=fl\left(\frac{fl(fl(11.0)-fl(\sqrt{117}))}{fl(2.0)}\right)=fl\left(\frac{fl(11.0-10.82)}{2.0}\right)=
fl\left(\frac{0.18}{2.0}\right)=0.09.
\end{gather*}

The relative errors are:
\begin{gather*}
\text{rel error in }r_{1}=\left|\frac{10.90832691-10.91}{10.90832691}\right|=1.5\times10^{-4}\\
\text{rel error in }r_{2}=\left|\frac{0.09167308680-0.09}{0.09167308680}\right|=1.8\times10^{-2}.
\end{gather*}
Notice the larger relative error in $r_{2}$ compared to that of $r_{1},$
about a factor of $100$, which is due to cancellation of leading
digits when we compute $11.0-10.82$.

One way to fix this problem is to rewrite the offending expression
by rationalizing the numerator:
\begin{equation*}
r_{2}=\frac{11.0-\sqrt{117}}{2}=\left(\frac{1}{2}\right) \frac{11.0-\sqrt{117}}{11.0+\sqrt{117}}\left(11.0+\sqrt{117}\right)=\left(\frac{1}{2}\right) \frac{4}{11.0+\sqrt{117}}=\frac{2}{11.0+\sqrt{117}}.
\end{equation*}
If we use this formula to compute $r_{2}$ we get:
\begin{equation*}
fl(r_{2})=fl\left(\frac{2.0}{fl(11.0+fl(\sqrt{117}))}\right)=fl\left(\frac{2.0}{21.82}\right)=0.09166.
\end{equation*}
The new relative error in $r_{2}$ is:
\begin{equation*}
\text{rel error in }r_{2}=\left(\frac{0.09167308680-0.09166}{0.09167308680}\right)=1.4\times10^{-4},
\end{equation*}
an improvement about a factor of $100,$ even though in the new
way of computing $r_{2}$ there are two operations where rounding error
happens instead of one.

\begin{example}\label{example:adding_floatingpt}
The simple procedure of adding numbers, even if they do not have mixed signs, can accumulate large errors due to rounding or chopping. Several sophisticated algorithms to add large lists of numbers with accumulated error smaller than straightforward addition exist in the literature (see, for example, Higham (1993)).

For a simple example, consider using four-digit arithmetic with rounding, and computing the average of two numbers, $\frac{a+b}{2}$. For $a=2.954$ and $b=100.9$, the true average is $51.927$. However, four-digit arithmetic with rounding yields:
\begin{equation*}
fl\left( \frac{100.9+2.954}{2} \right) = fl\left( \frac{fl(103.854)}{2}\right)=fl\left( \frac{103.9}{2}\right)=51.95,
\end{equation*}
which has a relative error of $4.43\times 10^{-4}$. If we rewrite the averaging formula as $a+\frac{b-a}{2}$, on the other hand, we obtain 51.93, which has a much smaller relative error of $5.78\times 10^{-5}$. The following table displays the exact and 4-digit computations, with the corresponding relative error at each step.

\begin{equation*}
\begin{array}{lcccccccll}
\hline
                 & a     & b    & a+b     & \frac{a+b}{2} & b-a     & \frac{b-a}{2} & a+\frac{b-a}{2} \\ \hline
4-\text{digit rounding} & 2.954  & 100.9 & 103.9   & 51.95         & 97.95   & 48.98         & 51.93           \\
\text{Exact}            &       &       & 103.854 & 51.927        & 97.946  & 48.973        & 51.927         \\
\text{Relative error}   &       &       & 4.43e-4 & 4.43e-4       & 4.08e-5 & 1.43e-4       & 5.78e-5         \\ \hline
\end{array}
\end{equation*}
\end{example}

\begin{example}
There are two standard formulas given in textbooks to compute the sample variance $s^2$ of the numbers $x_1,...,x_n$:

1. $s^2=\frac{1}{n-1} \left[ \sum_{i=1}^n x_i^2 - \frac{1}{n} \left( \sum_{i=1}^n x_i \right)^2 \right]$,
2. First compute $\bar{x}=\frac{1}{n} \sum_{i=1}^n x_i$, and then $s^2=\frac{1}{n-1} \sum_{i=1}^n (x_i-\bar{x})^2$.

Both formulas can suffer from roundoff errors due to adding large lists of numbers if $n$ is large, as mentioned in the previous example. However, the first formula is also prone to error due to cancellation of leading digits (see Chan et al. (1983) for details).

For an example, consider four-digit rounding arithmetic, and let the data be $1.253,2.411,3.174$. The sample variance from formula 1 and formula 2 are, 0.93 and 0.9355, respectively. The exact value, up to 6 digits, is 0.935562. Formula 2 is a numerically more stable choice for computing the variance than the first one.
\end{example}

\begin{example}
We have a sum to compute:

\begin{equation*}
e^{-7}=1+\frac{-7}{1}+\frac{(-7)^{2}}{2!}+\frac{(-7)^{3}}{3!}+\cdots+\frac{(-7)^{n}}{n!}+\cdots.
\end{equation*}
The alternating signs make this a potentially error prone calculation.

Python reports the "exact" value for $e^{-7}$ as 0.0009118819655545162. If we use Python to compute this sum with $n=20$, the result is 0.009183673977218275. Here is the Python code for the calculation:
\end{example}

In [ ]:
import numpy as np
from scipy.special import factorial

In [ ]:
sum = 1.0
for n in range(1, 21):
    sum += (-7)**n/factorial(n)
sum

This result has a relative error of 9.1. We can avoid this huge error if we simply rewrite the above sum as
\begin{equation*}
e^{-7}=\frac{1}{e^{7}}=\frac{1}{1+7+\frac{7^{2}}{2!}+\frac{7^{3}}{3!}+...}.
\end{equation*}
The Python code for this computation using $n=20$ is below:

In [ ]:
sum = 1.0
for n in range(1, 21):
    sum += 7**n/factorial(n)
sum = 1/sum
sum

The result is 0.0009118951837867185, which has a relative error of $1.4\times 10^{-5}$.

\begin{exercise}\label{ex:1.3-2}
The $x$-intercept of the line passing through the points $(x_{1},y_{1})$ and $(x_{2},y_{2})$ can be computed using either one of the following formulas:

\begin{equation*}
x=\frac{x_{1}y_{2}-x_{2}y_{1}}{y_{2}-y_{1}}
\end{equation*}
or,
\begin{equation*}
x=x_{1}-\frac{(x_{2}-x_{1})y_{1}}{y_{2}-y_{1}}
\end{equation*}
with the assumption $y_1 \neq y_2$.

a. Show that the formulas are equivalent to each other.

b. Compute the $x$-intercept using each formula when $(x_{1},y_{1})=(1.02,3.32)$ and $(x_{2},y_{2})=(1.31,4.31)$. Use three-digit rounding arithmetic.

c. Use Python (or a calculator) to compute the $x$-intercept using
the full-precision of the device (you can use either one of the formulas). Using this result, compute the relative and absolute errors of the
answers you gave in part (b). Discuss which formula is better and
why.
\end{exercise}

### Exercise 1.3-3
Write two functions in Python to compute the binomial coefficient $\binom{m}{k}$
using the following formulas:

a. $\binom{m}{k}=\frac{m!}{k!(m-k)!}$ ($m!$ is \textbf{scipy.special.factorial(m)} in Python.)

b. $\binom{m}{k}=(\frac{m}{k})(\frac{m-1}{k-1})\times...\times(\frac{m-k+1}{1})$

Then, experiment with various values for $m,k$ to see which formula
causes overflow first.

### Exercise 1.3-4
Polynomials can be evaluated in a nested form (also called Horner's method) that has two advantages:
the nested form has significantly less computation, and it can reduce
roundoff error. For

\begin{equation*}
p(x)=a_{0}+a_{1}x+a_{2}x^{2}+...+a_{n-1}x^{n-1}+a_{n}x^{n}
\end{equation*}
 its nested form is
\begin{equation*}
p(x)=a_{0}+x(a_{1}+x(a_{2}+...+x(a_{n-1}+x(a_{n}))...)).
\end{equation*}
 Consider the polynomial $p(x)=x^{2}+1.1x-2.8$.

a. Compute $p(3.5)$ using three-digit rounding, and three-digit chopping arithmetic.
What are the absolute errors? (Note that the exact value of $p(3.5)$
is 13.3.)

b. Write $x^{2}+1.1x-2.8$ in nested form by these simple steps:
\begin{equation*}
x^{2}+1.1x-2.8=(x^{2}+1.1x)-2.8=(x+1.1)x-2.8.
\end{equation*}
Then compute $p(3.5)$ using three-digit rounding and chopping using the
nested form. What are the absolute errors? Compare the errors with
the ones you found in (a).

\begin{exercise}\label{ex:1.3-5}
Consider the polynomial written in standard form: $5x^{4}+3x^{3}+4x^{2}+7x-5$.

a. Write the polynomial in its nested form. (See the previous problem.)

b. How many multiplications does the nested form require when we evaluate
the polynomial at a real number? How many multiplications does the
standard form require? Can you generalize your answer to any $n$th
degree polynomial?
\end{exercise}



#### Arya and the unexpected challenges of data analysis 

Meet Arya! Arya is a college student interested in math, biology, literature, and acting. Like a typical college student, she texts while walking on campus, complains about demanding professors, and argues in her blogs that homework should be outlawed.

Arya is taking a chemistry class, and she performs some experiments in the lab to find the weight of two substances. Due to difficulty in making precise measurements, she can only assess the weights to four-significant digits of accuracy: 2.312 grams and 0.003982 grams. Arya's professor wants to know the product of these weights, which will be used in a formula.

![](images/Arya_data_1.png)

Arya computes the product using her calculator: $2.312\times0.003982=0.009206384$, and stares at the result in bewilderment. The numbers she multiplied had four-significant digits, but the product has seven digits! Could this be the result of some magic, like a rabbit hopping out of a magician's hat that was only a handkerchief a moment ago? After some internal deliberations, Arya decides to report the answer to her professor as $0.009206$. Do you think Arya was correct in not reporting all of the digits of the product?

![](images/Arya_data_2.png)

### Sources of error in applied mathematics

Here is a list of potential sources of error when we solve a problem.

1. Error due to the simplifying assumptions made in the development of a mathematical model for the physical problem.
2. Programming errors.
3. Uncertainty in physical data: error in collecting and measuring data.
4. Machine errors: rounding/chopping, underflow, overflow, etc.
5. Mathematical truncation error: error that results from the use of numerical methods in solving a problem, such as evaluating a series by a finite sum, a definite integral by a numerical integration method, solving a differential equation by a numerical method.

### Example
The volume of the Earth could be computed using the formula for the volume of a sphere,
$V=4/3 \pi r^{3}$, where $r$ is the radius. This computation involves the following approximations:

1. The Earth is modeled as a sphere (modeling error)
2. Radius $r\approx6370$ km is based on empirical measurements (uncertainty in physical data)
3. All the numerical computations are done in a computer (machine error)
4. The value of $\pi$ has to be truncated (mathematical truncation error)

### Exercise
The following is from "Numerical mathematics and computing" by
Cheney & Kincaid (2012):

*In 1996, the Ariane 5 rocket launched by the European Space
Agency exploded 40 seconds after lift-off from Kourou, French Guiana.
An investigation determined that the horizontal velocity required
the conversion of a 64-bit floating-point number to a 16-bit signed
integer. It failed because the number was larger than 32,767, which
was the largest integer of this type that could be stored in memory.
The rocket and its cargo were valued at $500 million*.

Search online, or in the library, to find another example of computer
arithmetic going very wrong! Write a short paragraph explaining the
problem, and give a reference.

# References

Goldberg, D., 1991. What every computer scientist should know about floating-point arithmetic. ACM Computing Surveys (CSUR), 23(1), pp.5-48.

Higham, N.J., 1993. The accuracy of floating point summation. SIAM Journal on Scientific Computing, 14(4), pp.783-799.

Chan, T.F., Golub, G.H., and LeVeque, R.J., 1983. Algorithms for computing the sample variance: Analysis and recommendations. The American Statistician, 37(3), pp.242-247.

Cheney, E.W., and Kincaid, D.R., 2012. Numerical mathematics and computing. Cengage Learning.